aexpect/ops.py

```diff
@@ -0,0 +1,479 @@
+    :type session: ShellSession
+    :param str path: path to the filesystem entry to stat
+    :returns: standard output of stat(1) querying the size of ``path``
+    :rtype: int
+    """
+    return int(stat(session, path, fmt=r"%s"))
+
+
+###############################################################################
+# test(1)
+###############################################################################
+
+
+def test(session, path, flags):
+    """
+    Wrapper for ``test``.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: path to the filesystem entry to test
+    :param str flags: flags passed to test
+    :returns: status of test(1) on ``path`` as True or False
+    :rtype: bool
+    """
+    return session.cmd_status(f"test {flags} '{path}'") == 0
+
+
+def is_directory(session, path):
+    """
+    Check if a directory exists on a given remote host.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: path to the directory to check
+    :returns: whether the directory exists
+    :rtype: bool
+    """
+    return test(session, path, flags="-d")
+
+
+def is_regular_file(session, path):
+    """
+    Check if a regular file exists on a given remote host.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: path to the regular file to check
+    :returns: whether the regular file exists
+    :rtype: bool
+    """
+    return test(session, path, flags="-f")
+
+
+def is_nonzero_size_file(session, path):
+    """
+    Check if a regular file exists on a given remote host.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: path to the regular file to check
+    :returns: whether the regular file exists and is nonzero size
+    :rtype: bool
+    """
+    return test(session, path, flags="-s")
+
+
+###############################################################################
+# file ops
+###############################################################################
+
+
+def ls(session, path, quote_path=True, flags="-1UNq"):  # pylint: disable=C0103
+    """
+    Run `ls` in given directory through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: absolute or relative path to list
+    :param bool quote_path: whether to quote (escape) or not (expand) the path
+    :param str flags: raw flags to pass to ls command
+    :returns: names of files in dir (without path components)
+    :rtype: [str]
+
+    Just like :py:func:`os.listdir`, does not include file names starting with
+    dot (`'.'`)
+    """
+    cmd = f'ls {flags} {quote(path)}' if quote_path else f'ls {flags} {path}'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status == 2:      # probably just nothing found
```

Comment:
Ls reports `2` when any of the expressions fails (eg. due to permission reasons). I'm wondering whether silent return of empty list is the right solution. For example "ls /tmp/*" reports many files and several "permission denied" messages and returns 2, but this code reports empty list with no warning, no failure...

Reply:
Yes, I had the same concern and found some of our tests would rely on this status to conclude the path is empty. Will consult with the author and if they confirm I would also prefer to simply drop this strange handling and error out as everywhere else.

Reply:
Yes please, especially the fact that it ignores the output and silently returns empty list as if nothing happened smells like troubles to me.

Reply:
My still ongoing problem with this is how `ls` would treat glob expansion when there are no items in the expected list:
```
[root@jupiter intra2net-qa-deps]# ls /tmp/no-files*
ls: cannot access '/tmp/no-files*': No such file or directory
[root@jupiter intra2net-qa-deps]# echo $?
2
```
It would be much better if `ls` would not mix this status with as a "serious trouble":
```
   Exit status:
       0      if OK,

       1      if minor problems (e.g., cannot access subdirectory),

       2      if serious trouble (e.g., cannot access command-line argument).
```
In other words what one would expect when expanding a list of paths that is zero is to get an empty list back. Do you perhaps have any other suggestion on how to go around this?

Reply:
I think the only option we have here is to manually parse the stderr for "No such file or directory" even though it seems quite crude for a solution. Do you have any better idea?

Reply:
I might have found an option in bash like
```
$ shopt -s nullglob
$ echo "$(doc/xxxx*/abc/*)"
$ echo "${list[@]}" 

list=(doc/*)
echo "${list[@]}"
doc/api doc/api_create.py doc/api_entry.html doc/api_serve.sh doc/guides doc/README.md doc/references
```
but it depends on the shell type and is different for zsh, ksh93, etc.

Reply:
Not sure again but how about simply restoring the glob function in the module and not relying on `ls` there but on something simple like
```
$ echo doc/*
doc/api doc/api_create.py doc/api_entry.html doc/api_serve.sh doc/guides doc/README.md doc/references
$ echo doc/xxxx*/abc/*

```
?

Reply:
Hello @pevogam, I'm sorry for not noticing this pending request earlier. Now the function is called `ls` so I'd expect it to act as ls, ideally without corner cases, if possible. But you are the author and most certainly the biggest user, so I think it makes sense for you to define the rules here. I'm fine with anything systematic, ideally without exceptions (or just minor corner cases).

Reply:
I will look into this and do some testing in advance then possibly push a new pull request. Just FYI, this is the last open issue regarding the current merged that I am aware of and instead of opening a new issue I will simply push the corresponding pull request next, I have no other pending issues to register at aexpect on my side at least.

Reply:
Follow up on https://github.com/avocado-framework/aexpect/pull/113